services/spar/test-integration/Test/Spar/SCIMSpec.hs

```diff
@@ -0,0 +1,261 @@
+{-# LANGUAGE ConstraintKinds     #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE QuasiQuotes         #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections       #-}
+{-# LANGUAGE TypeApplications    #-}
+{-# LANGUAGE ViewPatterns        #-}
+{-# LANGUAGE LambdaCase          #-}
+
+module Test.Spar.SCIMSpec where
+
+import Bilge
+import Bilge.Assert
+import Brig.Types.User
+import Control.Lens
+import Control.Monad.IO.Class
+import Control.Monad.Reader
+import Data.ByteString.Conversion
+import Data.Id
+import Data.Monoid
+import Data.UUID as UUID hiding (null, fromByteString)
+import Data.Text (pack, unpack)
+import System.Random
+import Text.Read (readEither)
+import Util
+
+import qualified Web.SCIM.Class.User              as SCIM
+import qualified Web.SCIM.Class.Auth              as SCIM
+import qualified Web.SCIM.Schema.ListResponse     as SCIM
+import qualified Web.SCIM.Schema.Common           as SCIM
+import qualified Web.SCIM.Schema.Meta             as SCIM
+import qualified Web.SCIM.Filter                  as SCIM
+
+import qualified Web.SCIM.Schema.User             as SCIM.User
+
+import qualified Data.Aeson as Aeson
+
+
+spec :: SpecWith TestEnv
+spec = do
+    specUsers
+
+
+specUsers :: SpecWith TestEnv
+specUsers = describe "operations with users" $ do
+    describe "POST /Users" $ do
+        it "creates a user in an existing team" $ do
+            env <- ask
+            -- Create a user via SCIM
+            user <- randomUser
+            scimStoredUser <- createUser user
+            let userid = scimUserId scimStoredUser
+            -- Check that this user is present in Brig and that Brig's view
+            -- of the user matches SCIM's view of the user
+            brigUser <- fmap decodeBody' . call . get $
+                ( (env ^. teBrig)
+                . header "Z-User" (toByteString' userid)
+                . path "/self"
+                . expect2xx
+                )
+            scimStoredUser `userShouldMatch` brigUser
+
+    describe "GET /Users" $ do
+        it "lists all users in a team" $ do
+            env <- ask
+            -- Create a user via SCIM
+            user <- randomUser
+            storedUser <- createUser user
+            -- Get all users via SCIM
+            users <- listUsers Nothing
+            -- Check that the SCIM user is present
+            liftIO $ users `shouldContain` [storedUser]
+            -- Check that the (non-SCIM-provisioned) team owner is present
```

Comment:
That feels wrong.  Why would we want that?

Reply:
We don't want that at this stage, though we also don't necessarily want to _not_ have that. I see no usecase for hiding non-SCIM-provisioned users; do you?

In case you just object to it on the grounds of "we don't need it":

I'm somewhat sure that in the future we will need it, though, and I'm more than somewhat sure that we'll want two-way sync instead of one-way sync at some point. I'm okay with it being out of scope for now, but I still want some checks to be present so that half a year later we don't end up with an architecture (and implicit invariants all over the code) that prevent us from having two-way sync and so on.

For me, it's the same principle as "try to make your code reasonably generic even if you don't strictly need it", but applied to architecture.

Just to make it perfectly clear, I don't rule out the possibility that my heuristics are wrong. I'm just not willing to allow "we don't strictly need this now" be considered a hard-and-fast rule for not doing something.

Reply:
If it turns out that I have miscalculated the cost of supporting the usecase of "see non-SCIM-provisioned users via SCIM" (in the code, or in tests, etc), we can surely revisit it.